### PR TITLE
Allow to asynchronously generate PDFs

### PR DESCRIPTION
**Pull request by yhabteab**
The current implementation of `HeadlessChrome::toPdf()` always assumes that it controls the event loop instance, i.e. `HeadlessChrome` creates and starts the event loop manually. This may work for most use cases as they are mostly triggered via Icinga Web, but not if you want to generate PDFs using a daemon. Since our scheduler uses the same global event instance, it is unfavourable to call `Factory::create()` over again occasionally.

refs https://github.com/Icinga/icingaweb2-module-reporting/issues/229

### CODE REVIEW

File: vendor/iio/libmergepdf/tcpdi/fpdf_tpl.php
Comment (nilmerg) [inline vendor/iio/libmergepdf/tcpdi/fpdf_tpl.php:1]: Is there no release? Are these errors or just warnings from phpstan?

**Review by nilmerg — CHANGES_REQUESTED** [review]
I'm not really convinced that the usage of Promise\Deferred is correct. I don't know what's correct nor have I ever used one myself (as mentioned offline) but it looks wrong. Will look into this tomorrow to see if I'm right.

File: vendor/iio/libmergepdf/tcpdi/fpdf_tpl.php
Comment (yhabteab) [reply vendor/iio/libmergepdf/tcpdi/fpdf_tpl.php:1]: It's PHP deprecation errors, Phpstan doesn't scan vendor directories.

File: library/Pdfexport/HeadlessChrome.php [hunk 1]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
+    protected function toPdfCommon(): ExtendedPromiseInterface
+    {
+        $deferred = new Promise\Deferred();
+        Loop::futureTick(function () use ($deferred) {
+            switch (true) {
+                case $this->remote !== null:
+                    try {
+                        $result = $this->jsonVersion($this->remote[0], $this->remote[1]);
+                        if (is_array($result)) {
+                            $parts = Str::trimSplit('/', $result['webSocketDebuggerUrl']);
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:298]: Please use explode again, as discussed

File: library/Pdfexport/HeadlessChrome.php [hunk 2]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
+                                if (! empty($pdf)) {
+                                    $deferred->resolve($pdf);
+                                } else {
+                                    $deferred->reject(
+                                        new Exception(
+                                            'Received empty response or none at all from browser.'
+                                            . ' Please check the logs for further details.'
+                                        )
+                                    );
+                                }
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:313]: Please move this after the switch-case, as discussed

File: library/Pdfexport/HeadlessChrome.php [hunk 3]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
+                            });
+
+                            return;
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:316]: and break here

File: library/Pdfexport/HeadlessChrome.php [hunk 4]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
+                        }
 
-                $pdf = null;
-                $chrome->stderr->on('data', function ($chunk) use (&$pdf, $chrome, $loop, $killer) {
-                    Logger::debug('Caught browser output: %s', $chunk);
+                        Logger::warning('Failed to connect to remote chrome: %s (%s)', implode(':', $this->remote), $e);
+                        Logger::debug($e->getTraceAsString());
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:329]: Please don't log the stacktrace, seems irrelevant to me

File: library/Pdfexport/HeadlessChrome.php [hunk 5]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
-                $chrome->start($loop);
+    protected function toPdfCommon(): ExtendedPromiseInterface
+    {
+        $deferred = new Promise\Deferred();
+        Loop::futureTick(function () use ($deferred) {
+            switch (true) {
+                case $this->remote !== null:
+                    try {
+                        $result = $this->jsonVersion($this->remote[0], $this->remote[1]);
+                        if (is_array($result)) {
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:297]: If this isn't true, you have to handle this case and reject the promise. Remember to do this outside the try-block.

File: library/Pdfexport/HeadlessChrome.php [hunk 6]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
+                    $killer = Loop::addTimer(10, function (TimerInterface $timer) use ($chrome, $deferred) {
+                        $chrome->terminate(6); // SIGABRT
+
+                        $interval = $timer->getInterval();
+                        // Reject/resolve the promise on its own event tick not to block the current one.
+                        Loop::futureTick(function () use ($deferred, $interval) {
+                            $deferred->reject(
+                                new HeadlessChromeException(
+                                    sprintf(
+                                        'Browser timed out after %d seconds without the expected output',
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:367]: Log this, and reject the promise with `Received empty response or none at all from browser. Please check the logs for further details.`

File: library/Pdfexport/HeadlessChrome.php [hunk 7]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
+                                            )
+                                        );
+                                    }
+                                });
+                            } catch (Exception $e) {
+                                // Reject/resolve the promise on its own event tick not to block the current one.
+                                Loop::futureTick(function () use ($e, $deferred) {
+                                    $deferred->reject(
+                                        new HeadlessChromeException(
+                                            'Failed to print PDF. An error occurred: ',
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:405]: Log this, as discussed. Afterwards, `HeadlessChromeException` should be redundant and removed.

File: library/Pdfexport/HeadlessChrome.php [hunk 8]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
+                                    if (! empty($pdf)) {
+                                        $deferred->resolve($pdf);
+                                    } else {
+                                        $deferred->reject(
+                                            new Exception(
+                                                'Received empty response or none at all from browser.'
+                                                . ' Please check the logs for further details.'
+                                            )
+                                        );
+                                    }
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:398]: Move this outside the try-block, after chrome has been terminated

File: library/Pdfexport/HeadlessChrome.php [hunk 9]
```diff
@@ -247,106 +264,169 @@ public function fromHtml($html, $asFile = false)
...
-        if (empty($pdf)) {
-            throw new Exception(
-                'Received empty response or none at all from browser.'
-                . ' Please check the logs for further details.'
-            );
-        }
+                        // Browser is either timed out (after 10s) and the promise should have already been rejected,
+                        // or it is terminated using its terminate() method, in which case the promise is also already
+                        // resolved/rejected. So, we don't need to resolve/reject the promise here.
+                    });
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:425]: Remember to return here, to not to throw too early

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Pdfexport/HeadlessChrome.php [hunk 10]
```diff
@@ -240,111 +243,168 @@ public function fromHtml($html, $asFile = false)
     }
 
     /**
-     * Export to PDF
+     * Tries to asynchronously generate/print a PDF raw string from the configured browser.
```
Comment (nilmerg) [inline library/Pdfexport/HeadlessChrome.php:246]: That's a description. The first line in doc blocks should be a statement.

Also:

* it doesn't print (you document a method of a class, in which case print means echo)
* Since it doesn't print nor writes to disk, the fact that it results in a PDF *string* is self explaining
* I don't see a reason to document *here* that a browser is used. If at all, that's something for the class docblock

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]
Didn't test the asynchronous variant. I guess you did that thoroughly yourself. :sweat_smile:

**Review by sukhwinder33445 — COMMENTED** [review]
Please remove [this](0ad0f1077a90e) commit, it has already [merged](https://github.com/Icinga/icingaweb2-module-pdfexport/pull/68) as a Composer patch.
